[C10] PG observability hooks.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110307
* #109739
* __->__ #108815

Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

kumpera committed:
[C10] PG observability hooks.

Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

wconstab commented on test/distributed/test_hooks.py:
```diff
@@ -0,0 +1,65 @@
+        pg0 = dist.new_group(ranks=[0, 1])
+        pg1 = dist.new_group(ranks=[2, 3])
+
+        self.assertEqual(len(pgs), 2)
```
surprised this isn't 3?
also would be nice to exercise the 'pg_name' field in the test- what would it be by default?

kumpera replied:
It's not 3 because each rank only creates 2 PGs, we only report PGs that are created.
This is an idiosyncrasy of the SPMD model of PG creation we have.

wconstab replied:
oh, that does make sense (but is surprising too- mby worth a comment)

wconstab commented on torch/csrc/distributed/c10d/Backend.cpp:
```diff
@@ -15,4 +17,38 @@ void Backend::init() {
+
+void Backend::emitCollectiveEnd(const Work& work) {
+  details::EventInfo evt;
+  evt.event_kind = details::EventKind::CollectionEnd;
```
helper for creating event and filling common fields?

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

kumpera commented:
There are multiple test failures on this PR that I'm working on.

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

xw285cornell commented:
This looks great! qq, it seems it's now writing to an os.pipe instead of committing to TCPStore. Previously it was the rank0 aggregate info and then tell everyone who's the stragger (which makes rank0 the bottleneck). Wondering how do we gather the global info from the pipe of each rank?

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/hooks.py:
```diff
@@ -0,0 +1,138 @@
+def _c10d_pg_hooks_loops():
+    while True:
+        # we don't care about the result, this is how we implement notification
+        _ = os.read(_pp_r, 1)
```
Do we really need this mechanism? You could write a pure C++ wrapper around `std::deque` which could provide methods like `queue.poll(timeout)` and `queue.read()`

kumpera replied:
The original design was a wrapper around std::deque but using a blocking wait. 
This was problematic as it blocked python from shutting down - regardless if the thread was a daemon thread or not.


I tried using atexit callbacks to unstuck the background thread but that didn't work due to python shutdown behavior.

We could switch to using that design with a timed wait which hopefully would not happen and increase the number of context switches.

kumpera replied:
It might make sense to switch to std::deque + timed wait once we have a c++ alternative as the performance implication of the python API would not matter as much.

pritamdamania87 commented on torch/distributed/hooks.py:
```diff
@@ -0,0 +1,138 @@
+def _c10d_pg_hooks_loops():
+    while True:
+        # we don't care about the result, this is how we implement notification
+        _ = os.read(_pp_r, 1)
```
Do we block forever on `os.read` if there is nothing to read? If so, how does this thread shutdown cleanly?

kumpera replied:
cpython can interrupt `os.read`. This is the reason of the odd choice of pipes.

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -935,27 +935,34 @@ void ProcessGroupNCCL::ncclCommWatchdog() {
 
-void ProcessGroupNCCL::logWorkStart(WorkNCCL& work) {
-  if (work.startTraceUpdated_)
+void ProcessGroupNCCL::logWorkStart(WorkNCCL& work, bool emitDesyncInfo) {
```
ProcessGroupNCCL knows exactly when a collective error/timeout occurs. Should we add callbacks for those too?

wconstab replied:
are you referring to watchdog functionality? i think the point of these hooks is to start unbundling some of the watchdog functionality to another more reliable service, hence, not building further on top of that.  (apart from using it to call the hooks)

pritamdamania87 replied:
I meant the watchdog could just call the hooks when it detects a failure/timeout and all the heavy lifting around dealing with failures/timeouts could happen elsewhere as you suggest.

For example, the current start and end hooks can be used to detect a timeout but it won't tell you about errors. The watchdog uses [ncclCommGetAsyncError](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclcommgetasyncerror) to detect errors and it would be good to have visibility into those as well.

kumpera replied:
There are two more events that are worth surfacing.

Collective enqueueing, which would help detect some classes of deadlocks.
Collective error, this is really tricky since when errors happen we might have a deadlocked python but still important.

pritamdamania87 reviewed (commented):
This would be super useful, thanks a lot for working on this!

wconstab commented on torch/distributed/hooks.py:
```diff
@@ -0,0 +1,138 @@
+import logging
```
yea, this is a good point.  i think we can answer it in 2 parts.

first is that doing a background-thread in python here is more of a demonstration iiuc.  we'd suggest a more robust way to use these hooks is to have them initiate a fast, async request to a service to store some data and then return.  this way, all the critical functionality is happening in the remote service (in another process, or even machine)

second is that we could also support registering c++ hooks.  i think it would be trivial to add, following exactly from the python API and mirroring the 'CollectiveStatus' object as a c++ class.  We might do this but if you want it and want to send a PR that'd be great too.

kumpera commented on torch/distributed/hooks.py:
```diff
@@ -0,0 +1,138 @@
+    "CollectiveStatus",
+    "COLLECTIVE_HOOK_TYPE",
+    "PG_HOOK_TYPE",
+    "register_collective_start_hook",
```
Yes, I'll land collective failure callbacks and C++ support on top of this PR as both of them will require a more reviewing time on top of this one.

kumpera commented on torch/distributed/hooks.py:
```diff
@@ -0,0 +1,138 @@
+
+
+def _c10d_pg_hooks_loops():
+    while True:
```
There's no stop condition since there's currently no mechanism to disable event delivery.

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -935,27 +935,34 @@ void ProcessGroupNCCL::ncclCommWatchdog() {
+  if (terminateProcessGroup_.load())
     return;
 
   // In case the start of the work hasn't been logged
```
do we have cases where we would not have called 'emitCollectiveStart' for this work? i don't understand the reason for `!work.startTraceUpdated_`

kumpera commented:
> This looks great! qq, it seems it's now writing to an os.pipe instead of committing to TCPStore. Previously it was the rank0 aggregate info and then tell everyone who's the stragger (which makes rank0 the bottleneck). Wondering how do we gather the global info from the pipe of each rank?

The pipe is just how I implement an interruptible semaphore that can be shared by python and c++.

wconstab commented on torch/distributed/hooks.py:
```diff
@@ -0,0 +1,138 @@
+        try:
+            event_kind = evt.pop("event_kind", None)
+            if event_kind is None:
+                logger.warning(
```
should we perhaps count these warnings and exit the loop if they are repeating? avoid spamming?

kumpera replied:
Good point, I added two measures, one to always sleep once running into an internal failure and another to bail out after a fixed number of errors.

wconstab commented on torch/distributed/hooks.py:
```diff
@@ -0,0 +1,138 @@
+                        e,
+                    )
+        except Exception as e:
+            # We have to keep processing otherwise the queue will grown infinitely large
```
should we instead bound the size of the queue from the producer side? and issue a warning from the enqueue if it gets full?

kumpera replied:
I'm making the buffer bounded. good call, this avoid having to worry about the pipe being full too

wconstab commented on torch/distributed/hooks.py:
```diff
@@ -0,0 +1,138 @@
+                )
+                continue
+
+            if event_kind == 0:
```
probably worth it to properly define these event enums in C++ and bind them out.

kumpera replied:
done

wconstab commented on torch/csrc/distributed/c10d/Hooks.cpp:
```diff
@@ -0,0 +1,53 @@
+    return;
+
+  std::unique_lock<std::mutex> lock(event_queue_lock);
+  event_queue.push_back(std::move(evt));
```
should we implement a max size for this queue and warn if it's reached?  should we then disable event collection?

kumpera replied:
I added a max size and just collect failures instead.

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

wconstab commented on torch/distributed/hooks.py:
```diff
@@ -0,0 +1,174 @@
+            for cb in cb_list:
+                try:
+                    cb(status)
+                except Exception as e:
```
pros/cons of leaving cb exception outside of internal_failures counter?

kumpera replied:
Pros: when there are more than one registered callback, they won't interfere with each other (the first failing won't mess with the second).

Cons: This might be spammy.

wconstab commented on torch/csrc/distributed/c10d/Backend.cpp:
```diff
@@ -15,4 +32,26 @@ void Backend::init() {
+  try {
+    evt.duration_ms = work.getDuration();
+  } catch (std::exception& e) {
+    C10D_INFO("Duraction not available {}", e.what());
```
hmm. this is probably not ok-

our 'disabled by default' approach for this feature applies only to the inner call to `details::enqueue_c10d_event(std::move(evt));`

however, even if that's disabled, we're still gonna swallow any exception that's raised from work.getDuration().  My concern is we'll prevent this error from being surfaced to the existing watchdog.  OR will it be re-raised on any subsequent access to work obj?

kumpera replied:
There's a big todo for this part, it's due to an API design flaw, let me fix it.

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

kumpera committed:
Update on "[C10] PG observability hooks."


Expose a set of observability hooks into C10D such that our users can
detect collectives failure both faster and more easily.

The design is similar to NCCL desync debug that it minimized the
overhead by doing most of the work out of the main thread.

This PR introduces a new module torch.distributed.hooks that exposes the following set of methods:

    register_collective_start_hook
    register_collective_end_hook
    register_process_group_hook

The process group hook exposes PG creation on the member ranks and call them inline from the
the PG creation code. This is fine since this happens during initialization and a limited number of times.

The collective start/end hooks are fired from a single background thread. It reads
events from a C++ queue and dispatches over.

Queue notification is oddly done using a pipe, this is needed so python can abort the thread on shutdown
and have it as background thread. This is not possible with more reasonable choices like a condvar.

[ghstack-poisoned]

wconstab reviewed (approved):
I think this is good now! Thanks for getting it into good shape @kumpera.

fduwjj commented on torch/distributed/hooks.py:
```diff
@@ -0,0 +1,169 @@
+
+def _c10d_pg_hooks_loops():
+    internal_failures = 0
+    while True:
```
nit: Maybe we can add some documents explaining that we need to keep Python alive so that cpp watchdog can call the callback and do something in python. Not need to be done in this PR, though.

wconstab replied:
well that is not really true imo.  we should not design our watchdog on top of python hooks probably.  i am not sure if we will use python hooks for production or if we will use them for experimentation.  the later PR in the stack adds c++ hooks.  we probably should use those if we build a critical component on top of them.  We could still use python hooks for some telemtry i think.

fduwjj commented on torch/distributed/hooks.py:
```diff
@@ -0,0 +1,169 @@
+def _c10d_pg_hooks_loops():
+    internal_failures = 0
+    while True:
+        # we don't care about the result, this is how we implement notification
```
nit: notification from c++ to python.

fduwjj reviewed (approved):
thanks for making this happening. So this is not centralized solution right?

wconstab commented:
> thanks for making this happening. So this is not centralized solution right?

this is not really a 'solution' (it is not implementing a new watchdog or new debug features) - more of a 'platform' on which people can build stuff.  So together with the c++ hooks PR we should be able to decouple our work on core watchdog code from new experiments or features.

wconstab commented:
@pytorchbot merge -f "flaky rocm job only"

albanD commented:
@pytorchbot -h

albanD commented:
@pytorchbot revert -m "Add a new torch.distributed.hooks namespace but does not document it, test was added this morning" -c landrace
